### PR TITLE
Change behaviors of create client, connection, and channel commands.

### PR DESCRIPTION
**Pull request by dai1975**
- just exit with status 1 if config.json has id but that client/connection/channel is not exists.
 - just exit with status 0 if config.json has id and that client/connection/channel is already opened.

### CODE REVIEW

File: core/channel.go [hunk 1]
```diff
@@ -74,6 +81,55 @@ func CreateChannel(pathName string, src, dst *ProvableChain, to time.Duration) e
...
+	getState := func(pc *ProvableChain) (chantypes.State, error) {
+		if pc.Chain.Path().ChannelID == "" {
+			return chantypes.UNINITIALIZED, nil
+		}
+
+		latestHeight, err := pc.LatestHeight()
+		if err != nil {
+			return chantypes.UNINITIALIZED, err
+		}
+		queryHeight := clienttypes.NewHeight(latestHeight.GetRevisionNumber(), 0)
```
Comment (siburu) [inline core/channel.go:101]: In most cases, `RevisionHeight == 0` means the genesis height.
In this case you should simply use the latest height.

File: core/connection.go [hunk 2]
```diff
@@ -84,6 +90,55 @@ func CreateConnection(pathName string, src, dst *ProvableChain, to time.Duration
...
+	getState := func(pc *ProvableChain) (conntypes.State, error) {
+		if pc.Chain.Path().ConnectionID == "" {
+			return conntypes.UNINITIALIZED, nil
+		}
+
+		latestHeight, err := pc.LatestHeight()
+		if err != nil {
+			return conntypes.UNINITIALIZED, err
+		}
+		queryHeight := clienttypes.NewHeight(latestHeight.GetRevisionNumber(), 0)
```
Comment (siburu) [inline core/connection.go:110]: In most cases, `RevisionHeight == 0` means the genesis height.
In this case you should simply use the latest height.

**Review by siburu — COMMENTED** [review]
@dai1975 Thank you. I have put some comments. Please check them and fix the code.

File: core/client.go [hunk 3]
```diff
@@ -10,7 +11,55 @@ import (
...
+	srcID := src.Chain.Path().ClientID;
+	dstID := dst.Chain.Path().ClientID;
+
+	if srcID == "" && dstID == "" {
+		return true, nil
+	} else if srcID == "" && dstID != "" {
+		return false, fmt.Errorf("dst client id is given but src's is not: %s", dstID)
+	} else if srcID != "" && dstID == "" {
+		return false, fmt.Errorf("src client id is given but dst's is not: %s", srcID)
+	}
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/client.go:20]: fixed

File: core/client.go [hunk 4]
```diff
@@ -10,7 +11,55 @@ import (
...
+
+	getState := func(pc *ProvableChain) (*clienttypes.QueryClientStateResponse, error) {
+		latestHeight, err := pc.LatestHeight()
+		if err != nil {
+			return nil, err
+		}
+		height := clienttypes.NewHeight(latestHeight.GetRevisionNumber(), 0);
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/client.go]: fixed

File: core/connection.go [hunk 5]
```diff
@@ -84,6 +90,55 @@ func CreateConnection(pathName string, src, dst *ProvableChain, to time.Duration
...
+
+	if srcID != "" && srcState == conntypes.UNINITIALIZED {
+		return false, fmt.Errorf("src connection id is given but that connection is not exists: %s", srcID);
+	}
+	if dstID != "" && dstState == conntypes.UNINITIALIZED {
+		return false, fmt.Errorf("dst connection id is given but that connection is not exists: %s", dstID);
+	}
+
+	if srcState == conntypes.OPEN && dstState == conntypes.OPEN {
+		fmt.Printf("connections are already created: src=%s, dst=%s\n", srcID, dstID)
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/connection.go]: fixed

File: core/connection.go [hunk 6]
```diff
@@ -84,6 +90,55 @@ func CreateConnection(pathName string, src, dst *ProvableChain, to time.Duration
...
+		return false, srcErr
+	}
+
+	dstState, dstErr := getState(src)
+	if dstErr != nil {
+		return false, dstErr
+	}
+
+	if srcID != "" && srcState == conntypes.UNINITIALIZED {
+		return false, fmt.Errorf("src connection id is given but that connection is not exists: %s", srcID);
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/connection.go]: fixed

File: core/channel.go (hunk 1 above)
Comment (dai1975) [reply core/channel.go]: fixed

File: core/connection.go (hunk 2 above)
Comment (dai1975) [reply core/connection.go]: fixed

File: tests/cases/tm2tm/tests/Makefile [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	test-conn-fail-unexist \
+	test-conn-fail-already-created \
+	test-channel-success \
+	test-channel-fail-unexist \
+	test-channel-fail-already-created \
+
+.PHONY: help
+help:
+	@echo "$(MAKE) <target>"
+	@echo "targets are:"
```
[the remark this answers is not in the document]
Comment (dai1975) [reply tests/cases/tm2tm/tests/Makefile]: tests/Makefile is removed

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -80,6 +80,8 @@ jobs:
         working-directory: ./tests/scripts
         run: |
           ./load_docker_images $CACHE_DOCKER_TENDERMINT_DIR tendermint-chain0:latest tendermint-chain1:latest
+      - name: Insatall softwares used by test
```
Comment (siburu) [inline .github/workflows/test.yml:83]: s/Insatall/Install/

File: core/channel.go [hunk 9]
```diff
@@ -74,6 +80,54 @@ func CreateChannel(pathName string, src, dst *ProvableChain, to time.Duration) e
...
+
+	if srcID != "" && srcState == chantypes.UNINITIALIZED {
+		return false, fmt.Errorf("src channel id is given but that channel does not exist: %s", srcID);
+	}
+	if dstID != "" && dstState == chantypes.UNINITIALIZED {
+		return false, fmt.Errorf("dst channel id is given but that channel does not exist: %s", dstID);
+	}
+
+	if srcState == chantypes.OPEN && dstState == chantypes.OPEN {
+		logger.Warn(fmt.Sprintf("channels are already created: src=%s, dst=%s", srcID, dstID))
```
Comment (siburu) [inline core/channel.go:125]: Thank you for using the logger!
However please use structured logging style like shown below.
https://github.com/hyperledger-labs/yui-relayer/blob/v0.5.1/core/naive-strategy.go#L109-L116

File: core/channel.go [hunk 10]
```diff
@@ -74,6 +80,54 @@ func CreateChannel(pathName string, src, dst *ProvableChain, to time.Duration) e
...
+		}
+		return res.Channel.State, nil
+	}
+
+	srcState, srcErr := getState(src)
+	if srcErr != nil {
+		return false, srcErr
+	}
+
+	dstState, dstErr := getState(src)
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/channel.go:112]: fixed

File: .github/workflows/test.yml (hunk 8 above)
Comment (dai1975) [reply .github/workflows/test.yml]: fixed

File: core/client.go [hunk 11]
```diff
@@ -10,31 +11,80 @@ import (
...
+		latestHeight, err := pc.LatestHeight()
+		if err != nil {
+			return nil, err
+		}
+
+		ctx := NewQueryContext(context.TODO(), latestHeight)
+		return pc.QueryClientState(ctx)
+	}
+
+	srcState := (*clienttypes.QueryClientStateResponse)(nil)
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/client.go]: fixed

File: tests/cases/tm2tm/scripts/test-create-client-success-single [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+RLY_BINARY=${SCRIPT_DIR}/../../../../build/yrly
+RLY="${RLY_BINARY} --debug"
+CONFIG=$HOME/.yui-relayer/config/config.json
+
+cp $CONFIG $CONFIG.tmp
+
+cat $CONFIG.tmp \
+    | jq '.paths.ibc01.dst["client-id"] |= ""' \
+    > $CONFIG
+$RLY tx clients --src-height 2 ibc01
```
[the remark this answers is not in the document]
Comment (dai1975) [reply tests/cases/tm2tm/scripts/test-create-client-success-single:19]: done.

File: core/connection.go [hunk 13]
```diff
@@ -84,6 +90,54 @@ func CreateConnection(pathName string, src, dst *ProvableChain, to time.Duration
...
+		}
+		return res.Connection.State, nil
+	}
+
+	srcState, srcErr := getState(src)
+	if srcErr != nil {
+		return false, srcErr
+	}
+
+	dstState, dstErr := getState(src)
```
[the remark this answers is not in the document]
Comment (dai1975) [reply core/connection.go]: fixed

File: core/channel.go (hunk 9 above)
Comment (dai1975) [reply core/channel.go]: done be structured.

File: tests/cases/tm2tm/scripts/test-create-client-success-single (hunk 12 above)
[the remark this answers is not in the document]
Comment (siburu) [reply tests/cases/tm2tm/scripts/test-create-client-success-single:19]: @dai1975 The current test code checks that the dst ID has changed, but it should also check that the dst ID is not `""`.

[the remark this answers is not in the document]
Comment (dai1975) [reply tests/cases/tm2tm/scripts/test-create-client-success-single:19]: checks not an empty string

[the remark this answers is not in the document]
Comment (siburu) [reply tests/cases/tm2tm/scripts/test-create-client-success-single:19]: great fix.thank you.

**Review by siburu — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**